### DISCUSSION TITLE
Decorator-based integration with Langfuse

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#1009] [Ideas]
I've heard from many Langfuse users who'd like to see a decorator-based integration with Langfuse in addition to the current integrations and SDKs.

_Inspiration from the [OTel docs](https://opentelemetry.io/docs/languages/python/instrumentation/#creating-spans-with-decorators):_
```python
@tracer.start_as_current_span("do_work")
def do_work():
    print("doing some work...")
```

Advantages:
- derive start_time, end_time, input, output automatically
- derive hierarchy between observations (spans, events, generations)
- potentially parse I/O of common libs automatically (LLM providers, vectordbs, frameworks) as an alternative way to build integrations instead of replacing imports or natively hooking into a framework

This is not a work in progress at the moment, but something that @maxdeichmann and I really want to prioritize soon. Please share your ideas and feedback here.

We believe this will be highly effective for the Python SDK, but may not be as applicable for the JS/TS SDK.

**Comment by lshalon** [comment]
Attached in the gist is a paradigm that I've found helpful for logging tool info to a root span.

https://gist.github.com/lshalon/a87e099c893c8e786ade439acb890508

Something like this functionality is great, but if integrated into the library, I believe advantages could include:
1. Ease of integration with nested traces
2. Easier management of newer functionality (i.e. describing what datasets a trace relies on)

**Reply by marcklingen** [reply]
Thank you for sharing your implementation, this is super helpful.

Agree on both arguments, adding decorators is high value to make instrumenting custom applications that do not or only partially rely on frameworks much easier.

**Comment by AshisGhosh** [comment]
Just wanted to contribute a quick implementation I put together for my own use. Key change for discussion would be explicitly passing through trace parameters in the decorator explicitly and as attributes.

https://gist.github.com/AshisGhosh/5d1ab84aa54e965005b2dbb959e612ad

**Reply by marcklingen** [reply]
Thank you! cc @hassiebp 

Hassieb just got started on a decorator implementation here, maybe there is something interesting to learn here: https://github.com/langfuse/langfuse-python/pull/425

@AshisGhosh in case you have any comments, feel free to add them to the PR 🙏

**Comment by hassiebp** [comment]
👑🐍 **Decorator-based Python SDK now in alpha-release, please test & leave your feedback here** 😊

Hi all - we're happy to share with you an alpha-release of the decorators-based Python SDK. This will significantly reduce code overhead and complexity in adopting Langfuse in Python projects and abstract away the tedious managing of the nesting hierarchy of Langfuse traces and observations. 

Feel free to see for yourself and check it out - your feedback is incredibly valuable for us to get a stable and complete version out of the door as soon as possible. You can leave your feedback right here! 🙌🏾 

Here's the gist:

```python
from langfuse.decorators import langfuse_context, observe


@observe()
def span_inside_trace():
    print("Hello, from a span inside a trace!")


@observe()
def function_to_trace():
    print("Hello, from the parent trace!")
    span_inside_trace()

function_to_trace()

langfuse_context.flush()
```

Voilà! ✨ Langfuse will generate a trace with a nested span for you.

**Links**:
- [Alpha-Release 2.20.0a3](https://pypi.org/project/langfuse/2.20.2a3/)
- [Pull request](https://github.com/langfuse/langfuse-python/pull/425)
- [Docs (preview)](https://langfuse-docs-git-feat-adds-python-decorators-langfuse.vercel.app/docs/sdk/python_decorators)
- [API Reference (preview)](https://feat-add-tracing-decorators.langfuse-python.pages.dev/langfuse/decorators)

👀 @AshisGhosh @phenomagentx @lshalon - as you showed interest here in the past 😊

**Reply by AshisGhosh** [reply]
Went through the docs and API reference - looks great!

**Reply by hassiebp** [reply]
Thanks a lot for your feedback!

**Reply by marcklingen** [reply]
thanks again @lshalon for sharing so much feedback via email 🙏
